delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingStreamBehaviour.scala

```diff
@@ -0,0 +1,284 @@
+import monix.execution.Scheduler
+
+import java.util.UUID
+import scala.util.{Failure, Success}
+
+/**
+  * The view index behavior defined in order to manage the indexing stream
+  */
+object IndexingStreamBehaviour {
+
+  type IndexingStream[V] = (ViewIndex[V], ProjectionProgress[Unit]) => Task[Stream[Task, Unit]]
+
+  type ClearIndex = String => Task[Unit]
+
+  private val logger: Logger = Logger[IndexingStreamBehaviour.type]
+
+  /**
+    * Creates a behavior for a StreamSupervisor that manages the stream
+    */
+  def apply[V](
+      shard: ActorRef[ShardCommand],
+      project: ProjectRef,
+      id: Iri,
+      fetchView: (Iri, ProjectRef) => UIO[Option[ViewIndex[V]]],
+      buildStream: IndexingStream[V],
+      clearIndex: ClearIndex,
+      projection: Projection[Unit],
+      retryStrategy: RetryStrategy[Throwable]
+  )(implicit uuidF: UUIDF, sc: Scheduler): Behavior[IndexingViewCommand[V]] =
+    Behaviors.setup[IndexingViewCommand[V]] { context =>
+      Behaviors.withStash(100) { buffer =>
+        import context._
+
+        def streamName(rev: Long) = s"${project}_${id}_$rev"
+
+        // When the stream completes or receives an error, we stop the actor
+        def onFinalize(uuid: UUID) = UIO.delay(self ! StreamStopped(uuid))
+
+        def runStream(current: Option[(ViewIndex[V], StreamSwitch)]): Unit = {
+          val io: Task[IndexingViewCommand[V]] = fetchView(id, project).flatMap {
+            case None                                    =>
+              UIO.delay(
+                logger.info("View {} in project {} can't be found, the indexing will be passivated", id, project)
+              ) >> current.fold(Task.unit)(_._2.stop).as(ViewNotFound)
+            case Some(fetchView) if fetchView.deprecated =>
+              UIO.delay(
+                logger.info("View {} in project {} is deprecated, the indexing will be passivated", id, project)
+              ) >> current.fold(Task.unit)(_._2.stop).as(ViewIsDeprecated)
+            case Some(fetchView)                         =>
+              for {
+                progress <- projection.progress(fetchView.projectionId)
+                switch   <-
+                  current match {
+                    case None                                                             =>
+                      UIO.delay(
+                        logger
+                          .debug(
+                            "Index view {} with revision {} in project {} will start from offset {}",
+                            id,
+                            fetchView.rev,
+                            project,
+                            progress.offset
+                          )
+                      ) >>
+                        StreamSwitch.run(
+                          streamName(fetchView.rev),
+                          buildStream(fetchView, progress),
+                          retryStrategy,
+                          onFinalize,
+                          onFinalize
```

Comment:
On cancel we would probably prefer to try to restart the stream, instead of just passivating it, no?

Reply:
The Stop message does not passivate the actor, it gets restarted after a backoff

Reply:
Where is that happening? I see `Behaviors.stopped` when receiving the message `StreamStopped`.
And if the actor gets restarted that means it also gets restarted on manual cancelation, which is not what we probably want, right?

Reply:
Remember entities is taking responsibility for restarting the actor as long as passivation is not explicitly asked (which only happens when it is deprecated).

Reply:
To me there is a difference between
1. The stream will start/restart when the cluster sharded entity is requested again by somewhere within our app
2. The stream will start/restart immediately after a failure

I understand that when using cluster sharding, and requesting an entity, this one will get spawn if not already present. But it is not that what I'm talking about.

Reply:
There is a retry strategy to restart the stream and it does not restart the actor.
But when this one is not applicable, only then it escalates to the actor which gets restarted somewhere else after a backoff.
Which makes sense because, the akka node may not be healthy

Reply:
I just want to make sure, when a stop (stream interruption) is requested, actor does not try to run again the stream.

I don't anywhere where this is explicitly coded though (I might have missed it). In the cluster sharding documentation it says

> When rememberEntities is enabled, whenever a Shard is rebalanced onto another node or recovers after a crash, it will recreate all the entities which were previously running in that Shard.

> To permanently stop entities send a ClusterSharding.Passivate to the ActorRef[ShardCommand] that was passed in to the factory method when creating the entity. Otherwise, the entity will be automatically restarted after the entity restart backoff specified in the configuration.

So I guess you are sending a `Passivate` in that case? You mentioned you are sending a `Passivate` on deprecation

Reply:
Yes, when we don't want the entity to respawn automatically we have to use passivation, not the stop message.
But so far, there is no endpoint to passivate a stream, just to restart it.
So passivation only occurs when the fetched view is deprecated.

I will add the links in the akka documentation in the comments.